### ISSUE TITLE
testFlowControl hangs indefinitely on rhel-7-5

### ISSUE THREAD

**Issue by martinpitt** [cockpit-project/cockpit#9662]
Right now it's (almost or completely) impossible to get the rhel-7-5 test to green on the infrastructure. Tests end with "236 tests, 193 passed, 0 failed, 39 skipped, 3 to go", and the log always ends with "ok 3" with no error message what went wrong. [example 1](https://fedorapeople.org/groups/cockpit/logs/pull-9656-20180714-034949-37302e74-verify-rhel-7-5/log.html), [example 2](https://fedorapeople.org/groups/cockpit/logs/pull-9639-20180713-143030-d09eb577-verify-rhel-7-5/log.html)

**Comment by martinpitt** [comment]
I'm following the container log while the test is running. It is now stuck at "ok 3", no further error messages.
```
$ virsh list
 Id    Name                           State
----------------------------------------------------
 291   rhel-7-5-127.0.0.2-2701        running

$ ps aux|cat
USER         PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
user           1  0.0  0.0  15884  1856 ?        Ss   18:35   0:00 /bin/sh /usr/local/bin/cockpit-tests --publish $TEST_PUBLISH --verbose
user       16939  0.2  0.0 1859224 28836 ?       Sl   20:09   0:08 /usr/sbin/libvirtd --timeout=30
user       17231  0.0  0.0  10420   688 ?        S    20:11   0:00 /usr/bin/timeout 12h /bin/sh -c cd cockpit/; set -ex; PRIORITY=0008 bots/make-checkout --verbose --base=master pull/9655/head f6aac193bd6802ddbb4e3630184789f1b7c622f9 && TEST_NAME=pull-9655-20180714-201111 TEST_REVISION=f6aac193bd6802ddbb4e3630184789f1b7c622f9 bots/tests-invoke --rebase=master verify/rhel-7-5 pull/9655/head
user       17232  0.0  0.0  15752  1720 ?        S    20:11   0:00 /bin/sh -c cd cockpit/; set -ex; PRIORITY=0008 bots/make-checkout --verbose --base=master pull/9655/head f6aac193bd6802ddbb4e3630184789f1b7c622f9 && TEST_NAME=pull-9655-20180714-201111 TEST_REVISION=f6aac193bd6802ddbb4e3630184789f1b7c622f9 bots/tests-invoke --rebase=master verify/rhel-7-5 pull/9655/head
user       17247  0.0  0.0  82676 13524 ?        S    20:11   0:00 /usr/bin/python3 bots/tests-invoke --rebase=master verify/rhel-7-5 pull/9655/head
user       17255  0.0  0.0  54056  3700 ?        S    20:11   0:00 ssh -o ServerAliveInterval=30 sink -- python sink pull-9655-20180714-201111-f6aac193-verify-rhel-7-5
user       31074  0.0  0.0 208960  5528 ?        Sl   20:15   0:00 /usr/sbin/virtlogd --timeout=30
user       31171  0.0  0.0  10420   684 ?        S    20:17   0:00 timeout 120m /build/cockpit/bots/../test/verify/run-tests --jobs 8
user       31172  0.0  0.0 266356 22680 ?        S    20:17   0:01 /usr/bin/python3 /build/cockpit/bots/../test/verify/run-tests --jobs 8
user       33334  0.0  0.0  15884  2076 ?        Ss   20:18   0:00 /bin/sh
user       81359  0.0  0.0 273644 17732 ?        S    20:52   0:00 /usr/bin/python3 /build/cockpit/bots/../test/verify/run-tests --jobs 8
user       81389 10.2  0.4 6156932 531404 ?      Sl   20:52   0:20 /usr/bin/qemu-system-x86_64 -machine accel=kvm -name guest=rhel-7-5-127.0.0.2-2701,debug-threads=on -S -object secret,id=masterKey0,format=raw,file=/home/user/.config/libvirt/qemu/lib/domain-291-rhel-7-5-127.0.0.2-2/master-key.aes -machine pc-i440fx-2.10,accel=kvm,usb=off,dump-guest-core=off -cpu host -m 1024 -realtime mlock=off -smp 1,sockets=1,cores=1,threads=1 -uuid fec5ffa4-3d73-4be2-aefa-88b9c10f0456 -display none -no-user-config -nodefaults -chardev socket,id=charmonitor,path=/home/user/.config/libvirt/qemu/lib/domain-291-rhel-7-5-127.0.0.2-2/monitor.sock,server,nowait -mon chardev=charmonitor,id=monitor,mode=control -rtc base=utc -no-shutdown -boot strict=on -device piix3-usb-uhci,id=usb,bus=pci.0,addr=0x1.0x2 -device virtio-scsi-pci,id=scsi0,bus=pci.0,addr=0x2 -drive file=/build/tmp/run/e9barrc3.qcow2,format=qcow2,if=none,id=drive-virtio-disk0,serial=ROOT -device virtio-blk-pci,scsi=off,bus=pci.0,addr=0x3,drive=drive-virtio-disk0,id=virtio-disk0,bootindex=1 -drive file=/build/cockpit/bots/machine/cloud-init.iso,format=raw,if=none,id=drive-ide0-0-1,readonly=on -device ide-cd,bus=ide.0,unit=1,drive=drive-ide0-0-1,id=ide0-0-1 -chardev pty,id=charserial0 -device isa-serial,chardev=charserial0,id=serial0 -device virtio-balloon-pci,id=balloon0,bus=pci.0,addr=0x4 -object rng-random,id=objrng0,filename=/dev/urandom -device virtio-rng-pci,rng=objrng0,id=rng0,bus=pci.0,addr=0x5 -netdev socket,mcast=230.0.0.1:6000,id=mcast0 -device rtl8139,netdev=mcast0,mac=52:54:01:00:00:01,bus=pci.0,addr=0x0f -netdev user,id=base0,restrict=on,net=172.27.0.0/24,hostname=m1.cockpit.lan,hostfwd=tcp:127.0.0.2:2701-:22,hostfwd=tcp:127.0.0.2:9591-:9090 -device rtl8139,netdev=base0,bus=pci.0,addr=0x0e -msg timestamp=on
user       81914  0.0  0.0  54120  4120 ?        S    20:52   0:00 ssh -p 2701 -i /build/cockpit/bots/machine/identity -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o BatchMode=yes -M -o ControlPath=/tmp/ssh-%h-%p-%r-81359 -o LogLevel=ERROR -l root 127.0.0.2 /bin/bash -c 'echo READY; read a'
user       81920  0.6  0.0 1160152 106764 ?      Sl   20:52   0:01 /usr/bin/chromium-browser --enable-plugins --enable-extensions --enable-user-scripts --enable-printing --enable-sync --auto-ssl-client-auth --headless --disable-gpu --no-sandbox --disable-setuid-sandbox --disable-namespace-sandbox --disable-seccomp-filter-sandbox --disable-sandbox-denial-logging --disable-pushstate-throttle --window-size=1280x1200 --remote-debugging-port=9675 about:blank
user       81983  0.3  0.0 947524 68080 ?        S    20:52   0:00 /usr/lib64/chromium-browser/chromium-browser --type=zygote --disable-seccomp-filter-sandbox --no-sandbox --headless --headless
user       82119  0.0  0.0 1604580 33856 ?       Sl   20:52   0:00 /usr/lib64/chromium-browser/chromium-browser --type=renderer --disable-seccomp-filter-sandbox --no-sandbox --use-gl=swiftshader-webgl --disable-databases --disable-gpu-compositing --service-pipe-token=E72B6E1C1E5673E22F8E77D25CB5F644 --lang=en-US --headless --num-raster-threads=4 --enable-main-frame-before-activation --service-request-channel-token=E72B6E1C1E5673E22F8E77D25CB5F644 --renderer-client-id=2 --shared-files=v8_natives_data:100,v8_snapshot_data:101
user       82120  0.3  0.0 1004572 70988 ?       Sl   20:52   0:00 /usr/lib64/chromium-browser/chromium-browser --type=gpu-process --disable-seccomp-filter-sandbox --headless --no-sandbox --headless --gpu-preferences=KAAAAAAAAACAAACAAQAAAAAAAAAAAGAAEAAAAAAAAAAAAAAAAAAAAAgAAAAAAAAA --use-gl=swiftshader-webgl --override-use-software-gl-for-tests --headless --service-request-channel-token=4606FDCF69C026132CFEC1E29DCC471E
user       82128  0.1  0.0 1066664 27752 ?       Sl   20:52   0:00 node /build/cockpit/test/common/cdp-driver.js 9675
user       82144  0.7  0.0 1696244 98480 ?       Sl   20:52   0:01 /usr/lib64/chromium-browser/chromium-browser --type=renderer --disable-seccomp-filter-sandbox --no-sandbox --use-gl=swiftshader-webgl --disable-databases --disable-gpu-compositing --service-pipe-token=F1245613C492D1B7BA2271AA2254D5E2 --lang=en-US --headless --num-raster-threads=4 --enable-main-frame-before-activation --service-request-channel-token=F1245613C492D1B7BA2271AA2254D5E2 --renderer-client-id=4 --shared-files=v8_natives_data:100,v8_snapshot_data:101

$ pstree -p 31172
run-tests(31172)───run-tests(81359)─┬─chromium-browse(81920)─┬─chromium-browse(81983)─┬─chromium-browse(82119)─┬─{chromium-browse}(82121)
                                    │                        │                        │                        ├─{chromium-browse}(82122)
                                    │                        │                        │                        ├─{chromium-browse}(82123)
                                    │                        │                        │                        ├─{chromium-browse}(82124)
                                    │                        │                        │                        ├─{chromium-browse}(82125)
                                    │                        │                        │                        ├─{chromium-browse}(82126)
                                    │                        │                        │                        ├─{chromium-browse}(82127)
                                    │                        │                        │                        ├─{chromium-browse}(82134)
                                    │                        │                        │                        ├─{chromium-browse}(82135)
                                    │                        │                        │                        ├─{chromium-browse}(82136)
                                    │                        │                        │                        ├─{chromium-browse}(82137)
                                    │                        │                        │                        ├─{chromium-browse}(82138)
                                    │                        │                        │                        └─{chromium-browse}(82139)
                                    │                        │                        └─chromium-browse(82144)─┬─{chromium-browse}(82145)
                                    │                        │                                                 ├─{chromium-browse}(82146)
                                    │                        │                                                 ├─{chromium-browse}(82147)
                                    │                        │                                                 ├─{chromium-browse}(82149)
                                    │                        │                                                 ├─{chromium-browse}(82150)
                                    │                        │                                                 ├─{chromium-browse}(82151)
                                    │                        │                                                 ├─{chromium-browse}(82152)
                                    │                        │                                                 ├─{chromium-browse}(82153)
                                    │                        │                                                 ├─{chromium-browse}(82154)
                                    │                        │                                                 ├─{chromium-browse}(82155)
                                    │                        │                                                 ├─{chromium-browse}(82156)
                                    │                        │                                                 ├─{chromium-browse}(82157)
                                    │                        │                                                 ├─{chromium-browse}(82170)
                                    │                        │                                                 └─{chromium-browse}(82200)
                                    │                        ├─chromium-browse(82120)─┬─{chromium-browse}(82176)
                                    │                        │                        ├─{chromium-browse}(82177)
                                    │                        │                        ├─{chromium-browse}(82178)
                                    │                        │                        ├─{chromium-browse}(82179)
                                    │                        │                        ├─{chromium-browse}(82180)
                                    │                        │                        └─{chromium-browse}(82181)
                                    │                        ├─{chromium-browse}(81982)
                                    │                        ├─{chromium-browse}(82049)
                                    │                        ├─{chromium-browse}(82050)
                                    │                        ├─{chromium-browse}(82051)
                                    │                        ├─{chromium-browse}(82052)
                                    │                        ├─{chromium-browse}(82053)
                                    │                        ├─{chromium-browse}(82054)
                                    │                        ├─{chromium-browse}(82056)
                                    │                        ├─{chromium-browse}(82057)
                                    │                        ├─{chromium-browse}(82059)
                                    │                        ├─{chromium-browse}(82060)
                                    │                        ├─{chromium-browse}(82061)
                                    │                        ├─{chromium-browse}(82116)
                                    │                        └─{chromium-browse}(82118)
                                    ├─node(82128)─┬─{node}(82129)
                                    │             ├─{node}(82130)
                                    │             ├─{node}(82131)
                                    │             ├─{node}(82132)
                                    │             ├─{node}(82133)
                                    │             ├─{node}(82140)
                                    │             ├─{node}(82141)
                                    │             ├─{node}(82142)
                                    │             └─{node}(82143)
                                    └─ssh(81914)
```

`virsh console 291` works and gives me a login prompt, so the VM seems alright. Unfortunately it is not possible to see (from the above at least) which particular test is currently running. I tried to attach gdb to the test process, but even in a root shell in the container I can't do that (ptrace: Operation not permitted.). strace just says that the process is hanging in a `read()` from a pipe from a `cdp-driver.js` process.

I finally killed the test process,  which ended the cdp-driver.js process and the VM along with it, but it didn't advance the log at all. The main run-tests process (31172) has no further children now.

Killing the orphaned chromium process(es) finally ran tests-policy and gave me the remaining log:
```
/usr/lib64/python3.6/site-packages/sklearn/externals/six.py:7: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
  import imp
# ----------------------------------------------------------------------
# testExpire (check_accounts.TestAccounts)
#
[0714/205425.995604:ERROR:gpu_process_transport_factory.cc(1017)] Lost UI shared context.
ftruncate() failed: Permission denied
DevTools listening on ws://127.0.0.1:9959/devtools/browser/fc3f7003-6c30-4312-b23b-ee9a15d9bea3
[0714/205426.026406:ERROR:zygote_host_impl_linux.cc(267)] Failed to adjust OOM score of renderer with pid 84062: Permission denied (13)
[0714/205426.146760:ERROR:zygote_host_impl_linux.cc(267)] Failed to adjust OOM score of renderer with pid 84087: Permission denied (13)
ok 2 testExpire (check_accounts.TestAccounts) # duration: 33s
# ----------------------------------------------------------------------
# testBasic (check_accounts.TestAccounts)
#
[0714/205431.957471:ERROR:gpu_process_transport_factory.cc(1017)] Lost UI shared context.
ftruncate() failed: Permission denied
DevTools listening on ws://127.0.0.1:9892/devtools/browser/19ecee87-4fdb-427e-aa6e-3ef66aae74b8
[0714/205431.993593:ERROR:zygote_host_impl_linux.cc(267)] Failed to adjust OOM score of renderer with pid 84166: Permission denied (13)
[0714/205432.145707:ERROR:zygote_host_impl_linux.cc(267)] Failed to adjust OOM score of renderer with pid 84194: Permission denied (13)
> warning: chage: user 'anton' does not exist in /etc/passwd
> warning: Password quality check failed:
 The password is a palindrome
> warning: Password quality check failed:
 The password is a palindrome
> warning: Password quality check failed:
 The password is a palindrome
ok 1 testBasic (check_accounts.TestAccounts) # duration: 29s
# ----------------------------------------------------------------------
# testBasic (check_apps.TestApps)
#
[0714/205419.830029:ERROR:gpu_process_transport_factory.cc(1017)] Lost UI shared context.
ftruncate() failed: Permission denied
DevTools listening on ws://127.0.0.1:9389/devtools/browser/260be046-1b86-4e77-9d7f-aec97d0d0fca
[0714/205419.862823:ERROR:zygote_host_impl_linux.cc(267)] Failed to adjust OOM score of renderer with pid 83811: Permission denied (13)
[0714/205420.013953:ERROR:zygote_host_impl_linux.cc(267)] Failed to adjust OOM score of renderer with pid 83838: Permission denied (13)
ok 4 testBasic (check_apps.TestApps) # duration: 48s
# ----------------------------------------------------------------------
# testFlowControl (check_connection.TestConnection)
#
[0714/205258.713353:ERROR:gpu_process_transport_factory.cc(1017)] Lost UI shared context.
ftruncate() failed: Permission denied
DevTools listening on ws://127.0.0.1:9675/devtools/browser/197dbeba-b7b0-4cd7-8c8d-94eddae7acca
[0714/205258.752931:ERROR:zygote_host_impl_linux.cc(267)] Failed to adjust OOM score of renderer with pid 82119: Permission denied (13)
[0714/205258.897472:ERROR:zygote_host_impl_linux.cc(267)] Failed to adjust OOM score of renderer with pid 82144: Permission denied (13)
Connection to 127.0.0.2 closed by remote host.
# Flake likely: 0.833 (clustering). Please file issue if this not a flake
# Flake tracked https://api.github.com/repos/cockpit-project/cockpit/issues/7635 likelihood 1.000
# 1 TESTS FAILED [3416s on cockpit-tasks-45xvh]
+ continue
+ for i in $(seq 1 30)
+ git -C cockpit fetch origin
+ git -C cockpit reset --hard
```

So it's worth trying to locally reproduce these. My suspicion is at `testFlowControl()` which might need to get skipped or at least modified on rhel-7-5 as that uses an old cockpit. This might actually demonstrate one of the cases that PR #9642 addresses? For some reason that didn't run rhel-7-5, I'll trigger that now.

**Comment by martinpitt** [comment]
I can reproduce this locally:
```
-> ph_go("/playground/speed")
<- {'type': 'undefined'}
-> switch to frame None
-> wait: ph_is_present("iframe.container-frame[name='cockpit1:localhost/playground/speed'][data-loaded]")
<- {'type': 'undefined'}
-> wait: !ph_is_visible(".curtains-ct")
<- {'type': 'undefined'}
-> wait: ph_is_visible("iframe.container-frame[name='cockpit1:localhost/playground/speed']")
<- {'type': 'undefined'}
-> switch to frame cockpit1:localhost/playground/speed
-> wait: ph_is_present("body")
<- {'type': 'undefined'}
-> wait: ph_is_visible("body")
<- {'type': 'undefined'}
-> wait: !ph_text_is("#pid","")
<- {'type': 'undefined'}
-> ph_text("#pid")
<- {'type': 'string', 'value': '1489'}
-> ph_set_val("#read-path","/dev/vda")
<- {'type': 'undefined'}
-> ph_click("#read-sideband",false)
<- {'type': 'undefined'}
-> wait: !ph_text_is("#speed","")
```
For some reason the `b.wait_text_not("#speed", "")` never times out. When running this test in an interactive browser, I get an interesting error message:

![test](https://user-images.githubusercontent.com/200109/42732567-4000ff06-8824-11e8-91ae-90c084f63b29.png)

When I click on "OK", the test resumes, the speed remains at 0 Bytes/s, and the test eventually fails on the unexpected message `/dev/vda: not a regular file`. I. e. it "almost succeeds".

There are several things wrong here. I will post a PR to skip the test on rhel-7-5 as it is not meant to succeed there, but I think we should also fix the underlying reason for the eternal hang.

@stefwalter , does that ring a bell?

**Comment by martinpitt** [comment]
PR #9664 disables the test on rhel-7-5 as a quick-fix to unbreak our PRs. It's not a complete solution to this, though.

**Comment by martinpitt** [comment]
When this happens this seems to kill all timers. In `ph_wait_cond()`, both the step and the timeout timer just run once, then they never fire again. This is the reason for the indefinite hang. Could the high sidechannel saturation do that?

**Comment by martinpitt** [comment]
Fixed in #9664
